node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,172 @@
+                .field(&error)
+                .finish(),
+            ChunkingError::SerializationError(error) => {
+                f.debug_tuple("SerializationError").field(&error).finish()
+            }
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of [`ValueOrChunk::Value`] if data size is less than or equal to
+    /// [`ChunkWithProof::CHUNK_SIZE_BYTES`] or a [`ValueOrChunk::ChunkWithProof`] if it is greater.
+    /// In the latter case it will return only the `chunk_index`-th chunk of the value's byte
+    /// representation.
+    pub fn new(data: V, chunk_index: u64) -> Result<Self, ChunkingError>
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::to_bytes(&data).map_err(ChunkingError::SerializationError)?;
+        // NOTE: Cannot accept the chunk size bytes as an argument without changing the
+        // IndexedMerkleProof. The chunk size there is hardcoded and will be used when
+        // determining the chunk.
+        if bytes.len() <= ChunkWithProof::CHUNK_SIZE_BYTES {
+            Ok(ValueOrChunk::Value(data))
+        } else {
+            let chunk_with_proof = ChunkWithProof::new(&bytes, chunk_index)
+                .map_err(ChunkingError::MerkleConstruction)?;
+            Ok(ValueOrChunk::ChunkWithProof(chunk_with_proof))
+        }
+    }
+}
+
+/// Represents an enum that can contain either a whole trie or a chunk of it.
+pub type TrieOrChunk = ValueOrChunk<TrieRaw>;
+
+impl<V: Debug> Display for ValueOrChunk<V> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            ValueOrChunk::Value(data) => f.debug_tuple("Value").field(data).finish(),
+            ValueOrChunk::ChunkWithProof(chunk) => f
+                .debug_struct("ChunkWithProof")
+                .field("index", &chunk.proof().index())
+                .field("hash", &chunk.proof().root_hash())
+                .finish(),
+        }
+    }
+}
+
+/// Represents the ID of a `TrieOrChunk` - containing the index and the root hash.
+/// The root hash is the hash of the trie node as a whole.
+/// The index is the index of a chunk if the node's size is too large and requires chunking. For
+/// small nodes, it's always 0.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkId(pub u64, pub Digest);
+
+impl TrieOrChunkId {
+    /// Returns the trie key part of the ID.
+    pub fn digest(&self) -> &Digest {
+        &self.1
+    }
+
+    /// Given a serialized ID, deserializes it for display purposes.
+    fn fmt_serialized(f: &mut Formatter, serialized_id: &[u8]) -> fmt::Result {
+        match bincode::deserialize::<Self>(serialized_id) {
+            Ok(ref trie_or_chunk_id) => Display::fmt(trie_or_chunk_id, f),
+            Err(_) => f.write_str("<invalid>"),
+        }
+    }
+}
+
+/// Helper struct to on-demand deserialize a trie or chunk ID for display purposes.
+pub struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
```

Comment:
Same as for `TrieOrChunkId`.

Reply:
Resolving because there apparently is a good reason for not doing that (https://github.com/casper-network/casper-node/pull/3255#discussion_r953816573)